src/sentry/api/authentication.py

```diff
@@ -232,3 +242,29 @@ def authenticate_credentials(self, request: Request, token):
             scope.set_tag("api_project_key", key.id)
 
         return (AnonymousUser(), key)
+
+
+FORWARDED_SSL_CERT_HEADER_NAME = "ssl-client-cert"
+
+
+class MutualTlsAuthentication(BaseAuthentication):
```

Comment:
This just accepts any certificate and does not perform any validation. Is there a description somewhere about how cross region communication should work and why certificates were picked? Note that Relay <-> Sentry already use an authentication system for similar purposes that uses public/key crypto based on ed25519 which signs the request bodies: https://github.com/getsentry/sentry/blob/master/src/sentry/api/authentication.py#L116C13-L121

Reply:
Previously [the plan was to use asymetric keys](https://www.notion.so/sentry/Cross-Region-Requests-Authentication-1859fc4c68154f31a24577bc9bc23932) similar to relay so that we don't have multiple kinds of server-server authentication.

Reply:
Correct -- in my proposal, this would occur at ingress alongside other TLS.  A small change in k8s ingestion allows peer certificates to be validated and their identity forward.  So in this world, sentry services do not need themselves to validate peer certificates.  Some explanation here: https://www.notion.so/sentry/Region-Internal-RPC-and-Mutual-TLS-Auth-ccc3935962e341eeb4094f1a66204c08

Reply:
@mitsuhiko So that approach is a bit different from what I had in mind.  For instance, what is the process of adding a new relay?  It seems that sentry is required to keep a registry of all public keys correct?  And furthermore, that configuration must be pushed out for any new relay?

Mutual TLS avoids this issue.  once we're going multi regional, it gets expensive to be pushing out public key infrastructure without a chain.  Using something like mutual TLS, we can use a root certificate as the means to deliver new private keys, and to have that validation work at infrastructure level (ingress) rather than application level.

From a performance level as well, doing this sort of crypto on high traffic can be expensive in python, no?

I don't know relay too well but I'm curious to know more about your experience deploying and using this system and what deploying new keys looks like, and how that would work in a multi regional world.

Reply:
Basically, I'm not convinced that relay's approach will work for hybrid cloud, or at least has some downsides worth discussing.

Reply:
In any case, I've begun discussions with @rgibert and got his perspective, I'd love to hear your perspective as well.  In the end, if we do want to re use the relay approach, I'm curious to know what our story is for generalizing it and how it works when deploying the secrets across regions.  Historically, I've seen mutual tls solve this problem with cert chains, but if that doesn't fit well in the envisioned operational story that's cool too.  I'd like to discuss and get some specifics on the desired state here since we don't have a generalized solution at hand yet.

Reply:
https://www.notion.so/sentry/Region-Internal-RPC-and-Mutual-TLS-Auth-ccc3935962e341eeb4094f1a66204c08

Reply:
> Asymmetric keys will require more complex configuration management as each region needs the control public key, and the region's private key, while control needs its private key and the public key of every region.
> 
> With mutual TLS how would we revoke a certificate for a single region?

The advantages are that the solutions here are generalizable.  For instance, it's possible to have the same root public key in a chain for relay AND rpc AND any other internal service.  you don't need each individual public key, only a single chain that can authenticate and authorize many (indefinitely expanding) sub services.  As opposed to symmetric keys which still must be shared for each individual unrelated object.

Revoking a certificate doesn't have to be hard -- simply revoke the subject name associated with the certificate from services, and force a new subject name / new ticket pairing.  This would be similar to rotation of certificates, too.

It is true that asymmetric keys do require some complexity, but it solves a lot of other issues around public key management, and it's more generalizable.  Even if I want to use the relay authentication, *I have to write new logic and code*.

1.  I need a new config store for region public keys, which must be synchronized somehow (the existing one is generalized)
2. I need to decide on the signature protocol and implement it in python (relay clients are in rust right now, right?)

I can definitely do that, to be clear.  I'm 100% accepting a hard decision and will implement it however we agree.  But I wanted to bring these thoughts up, at the very least, to consider the scale of the inter rpc authentication problem in a more general sense.

Reply:
The other beauty is the concept of subject names -- they are general string identities associated with the key itself.  You can store all kinds of information in these subject names, they don't have to be actually routable DSNs.

You can even use a schema like `region.service_name.version` and have various acl / rotation properties built in.

Reply:
> Revoking a certificate doesn't have to be hard -- simply revoke the subject name associated with the certificate from services, and force a new subject name / new private cut from the CA. 

It looks like both k8s ingest and nginx also have a 'crl' file that can be populated with revoked client certificates. That covers my concerns on how we remove access to malicious regions.

Reply:
I think this all makes a lot of sense. For as long as we document this clearly (ideally where the actual permission checks are happening) I'm happy. I have some anxiety reviewing permission related code that is unclear.